File: 3414-maximum-score-of-non-overlapping-intervals/3414-maximum-score-of-non-overlapping-intervals.cpp
class Solution {
public:
    using ll = long long;

    vector<int> maximumWeight(vector<vector<int>>& intervals) {
        int n = intervals.size();

        // Store: [left, right, weight, original_index]
        vector<array<ll, 4>> a(n);

        for (int i = 0; i < n; ++i) {
            a[i] = {
                intervals[i][0],
                intervals[i][1],
                intervals[i][2],
                i
            };
        }

        // Sort by left endpoint, then by original index
        sort(a.begin(), a.end(), [](const auto& x, const auto& y) {
            if (x[0] != y[0]) return x[0] < y[0];
            return x[3] < y[3];
        });

        // next[i] = first interval whose left > a[i].right
        // Strictly greater because touching intervals overlap.
        vector<int> next(n);

        vector<ll> starts(n);
        for (int i = 0; i < n; ++i)
            starts[i] = a[i][0];

        for (int i = 0; i < n; ++i) {
            next[i] = upper_bound(
                starts.begin(),
                starts.end(),
                a[i][1]
            ) - starts.begin();
        }

        /*
            dp[i][k] = best answer using intervals from i onward,
                       choosing at most k intervals.

            Each state stores:
            - maximum total weight
            - lexicographically smallest list of original indices
        */

        struct State {
            ll score = 0;
            vector<int> indices;
        };

        vector<vector<State>> dp(n + 1, vector<State>(5));

        // Compare two states:
        // 1. Higher score is better.
        // 2. If score is equal, lexicographically smaller indices are better.
        auto better = [](const State& x, const State& y) {
            if (x.score != y.score)
                return x.score > y.score;

            return lexicographical_compare(
                x.indices.begin(), x.indices.end(),
                y.indices.begin(), y.indices.end()
            );
        };

        for (int i = n - 1; i >= 0; --i) {

            // k = maximum number of intervals we may still choose
            for (int k = 1; k <= 4; ++k) {

                // Option 1: Skip current interval
                State skip = dp[i + 1][k];

                // Option 2: Take current interval
                State take;
                take.score = a[i][2] + dp[next[i]][k - 1].score;
                take.indices.push_back((int)a[i][3]);

                // The final answer must be sorted by index.
                for (int idx : dp[next[i]][k - 1].indices)
                    take.indices.push_back(idx);

                sort(take.indices.begin(), take.indices.end());

                dp[i][k] = better(take, skip) ? take : skip;
            }
        }

        return dp[0][4].indices;
    }
};

File: 3414-maximum-score-of-non-overlapping-intervals/3414-maximum-score-of-non-overlapping-intervals.java
import java.util.*;

class Solution {

    static class State {
        long score;
        int[] ids;

        State(long score, int[] ids) {
            this.score = score;
            this.ids = ids;
        }
    }

    public int[] maximumWeight(List<List<Integer>> intervals) {

        int n = intervals.size();

        // [left, right, weight, originalIndex]
        int[][] a = new int[n][4];

        for (int i = 0; i < n; i++) {
            a[i][0] = intervals.get(i).get(0);
            a[i][1] = intervals.get(i).get(1);
            a[i][2] = intervals.get(i).get(2);
            a[i][3] = i;
        }

        // Sort by right endpoint
        Arrays.sort(a, (x, y) -> {
            if (x[1] != y[1]) {
                return Integer.compare(x[1], y[1]);
            }
            return Integer.compare(x[3], y[3]);
        });

        // Right endpoints of sorted intervals
        int[] ends = new int[n];

        for (int i = 0; i < n; i++) {
            ends[i] = a[i][1];
        }

        /*
         * prev[i] = number of intervals before i
         * whose right endpoint is strictly less than
         * the left endpoint of interval i.
         */
        int[] prev = new int[n];

        for (int i = 0; i < n; i++) {
            prev[i] = lowerBound(ends, a[i][0]);
        }

        /*
         * dp[i][k]:
         * Best result using first i intervals
         * and choosing at most k intervals.
         */
        State[][] dp = new State[n + 1][5];

        // Empty selection
        for (int i = 0; i <= n; i++) {
            dp[i][0] = new State(0L, new int[0]);
        }

        for (int k = 1; k <= 4; k++) {
            dp[0][k] = new State(0L, new int[0]);
        }

        for (int i = 1; i <= n; i++) {

            int cur = i - 1;

            for (int k = 1; k <= 4; k++) {

                // Option 1: skip current interval
                State skip = dp[i - 1][k];

                // Option 2: take current interval
                State before = dp[prev[cur]][k - 1];

                int[] ids = addIndex(
                    before.ids,
                    a[cur][3]
                );

                State take = new State(
                    before.score + a[cur][2],
                    ids
                );

                dp[i][k] = better(take, skip)
                        ? take
                        : skip;
            }
        }

        return dp[n][4].ids;
    }

    /*
     * First position where arr[position] >= target.
     *
     * Therefore all positions before it satisfy:
     *
     * arr[position] < target
     */
    private int lowerBound(int[] arr, int target) {

        int left = 0;
        int right = arr.length;

        while (left < right) {

            int mid = left + (right - left) / 2;

            if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left;
    }

    /*
     * Adds an index while keeping the array sorted.
     *
     * There can be at most 4 indices,
     * so this operation is effectively O(1).
     */
    private int[] addIndex(int[] arr, int x) {

        int[] result = new int[arr.length + 1];

        int i = 0;

        while (i < arr.length && arr[i] < x) {
            result[i] = arr[i];
            i++;
        }

        result[i] = x;

        while (i < arr.length) {
            result[i + 1] = arr[i];
            i++;
        }

        return result;
    }

    /*
     * Returns true if a is better than b.
     *
     * First maximize score.
     * If scores are equal, choose lexicographically
     * smaller index array.
     */
    private boolean better(State a, State b) {

        if (a.score != b.score) {
            return a.score > b.score;
        }

        return lexicographicallySmaller(a.ids, b.ids);
    }

    /*
     * Compare two sorted arrays lexicographically.
     */
    private boolean lexicographicallySmaller(int[] a, int[] b) {

        int len = Math.min(a.length, b.length);

        for (int i = 0; i < len; i++) {

            if (a[i] != b[i]) {
                return a[i] < b[i];
            }
        }

        return a.length < b.length;
    }
}